### PR TITLE
Allow analyzers to be configured

### PR DESCRIPTION
**Pull request by K-Phoen**
I implemented a simple way to make analyzers configurable.

They can define their own configuration, that will be merged in the global one. This configuration is used to determine if the analyzer pass is enabled or not. When the analyzer is instantiated, its configuration is injected through the constructor.

I also added a new command to dump the default configuration: `./bin/phpsa config:dump-reference`

WDYT?

### CODE REVIEW

**Comment by ddmler** [comment]
@K-Phoen 
It seems you did not make ArrayShortDefinition and a few others configurable (don't have the trait)?

Can you also add a setting which would overwrite individual analyzers settings?
For example the setting min-php-version: 5.3 should overwrite ArrayShortDefinition with false even when this analyzer is set to true. I think such a setting would be quite nice for people who don't want to read documentation for all analyzers. (see #123 )

Also I think the default configuration should be in our .yml file. Then someone only has to open this and change something to false instead of searching what key to use first.

**Comment by K-Phoen** [comment]
> It seems you did not make ArrayShortDefinition and a few others configurable (don't have the trait)?

You're right, I didn't. I thought that the PR was already big enough and that I could make them configurable later :)

> Can you also add a setting which would overwrite individual analyzers settings?

That could be nice, I'll try to do that.

> Also I think the default configuration should be in our .yml file

Okay, I'll update it.

> (I guess your PR resolves #39 completely?)

It should \o/

File: src/Analyzer/Helper/ConfigurablePassTrait.php [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+ * @author Kévin Gomez https://github.com/K-Phoen <contact@kevingomez.fr>
+ */
+
+namespace PHPSA\Analyzer\Helper;
+
+use Symfony\Component\Config\Definition\Builder\TreeBuilder;
+
+trait ConfigurablePassTrait
+{
+    /**
```
Comment (ovr) [inline src/Analyzer/Helper/ConfigurablePassTrait.php]: Whole methods of this trait should not be `static` methods

File: src/Analyzer/Helper/ConfigurablePassTrait.php [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+     * {@inheritdoc}
+     */
+    public static function getName()
+    {
+        $fqcnParts = explode('\\', __CLASS__);
```
Comment (ovr) [inline src/Analyzer/Helper/ConfigurablePassTrait.php]: maybe get_class will be better? please write your opinion about it

File: src/Analyzer/Helper/ConfigurablePassTrait.php [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+    /**
+     * @param string $string
+     * @return string
+     */
+    private static function convertToSnakeCase($string)
+    {
+        $snakeCasedName = '';
+        $len = strlen($string);
+        for ($i = 0; $i < $len; ++$i) {
```
Comment (ovr) [inline src/Analyzer/Helper/ConfigurablePassTrait.php]: Space line before cycle

File: src/Analyzer/Helper/ConfigurablePassTrait.php [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        ;
+    }
+
+    /**
+     * @param string $string
+     * @return string
+     */
+    private static function convertToSnakeCase($string)
+    {
+        $snakeCasedName = '';
```
Comment (ovr) [inline src/Analyzer/Helper/ConfigurablePassTrait.php]: I suggest to name this variable like `$result`

File: src/Analyzer/Pass/ConfigurablePassInterface.php [hunk 5]
```diff
@@ -5,12 +5,15 @@
...
-     * @return TreeBuilder
+     * @return \Symfony\Component\Config\Definition\Builder\NodeDefinition
+     */
+    public static function getConfiguration();
+
+    /**
+     * @return string
      */
-    public function getConfiguration();
+    public static function getName();
```
Comment (ovr) [inline src/Analyzer/Pass/ConfigurablePassInterface.php]: Not static

File: src/Analyzer/Pass/Expression/ArrayShortDefinition.php [hunk 6]
```diff
@@ -33,7 +32,7 @@ public function pass(Expr\Array_ $expr, Context $context)
     }
 
     /**
-     * @return TreeBuilder
+     * @return array
```
Comment (ovr) [inline src/Analyzer/Pass/Expression/ArrayShortDefinition.php]: Good catch :smile_cat: :+1:

File: src/Command/DumpReferenceCommand.php [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        $analyzerConfiguration = Analyzer\Factory::getPassesConfigurations();
+        $configuration = new Configuration([], $analyzerConfiguration);
+
+        $dumper = new YamlReferenceDumper();
+        $output->writeln($dumper->dumpNode($configuration->getConfigTreeBuilder($analyzerConfiguration)->buildTree()));
```
Comment (ovr) [inline src/Command/DumpReferenceCommand.php]: Multi line fcall, will be better to split each call on own line

```
$output->writeln(
    $dumper->dumpNode(
        $configuration->getConfigTreeBuilder($analyzerConfiguration)->buildTree()
    )
);
```

File: src/Configuration.php [hunk 8]
```diff
@@ -22,32 +21,35 @@ class Configuration implements ConfigurationInterface
      * Create a configuration from array.
      *
      * @param array $configuration
+     * @param array $analyzersConfiguration
      */
-    public function __construct(array $configuration = [])
+    public function __construct(array $configuration = [], $analyzersConfiguration = [])
```
Comment (ovr) [inline src/Configuration.php]: array strict needed!

File: src/Configuration.php [hunk 9]
```diff
@@ -85,4 +98,36 @@ public function valueIsTrue($key)
     {
         return (bool) $this->configuration[$key];
     }
+
```
Comment (ovr) [inline src/Configuration.php]: I suggest to remove `\ArrayAccess`, will be better to protect any magic in Analyzer
Lets use `$this->configuration`

File: src/Analyzer/Factory.php [hunk 10]
```diff
@@ -6,49 +6,127 @@
 namespace PHPSA\Analyzer;
 
 use PHPSA\Analyzer;
+use PHPSA\Configuration;
+use Symfony\Component\Config\Definition\Builder\NodeDefinition;
 use Webiny\Component\EventManager\EventManager;
 use PHPSA\Analyzer\Pass as AnalyzerPass;
 
 class Factory
```
Comment (ovr) [inline src/Analyzer/Factory.php]: Please protect whole static methods, I dont think that static methods is needed
Thanks

File: src/Analyzer/Pass/ConfigurablePassInterface.php (hunk 5 above)
Comment (K-Phoen) [inline src/Analyzer/Pass/ConfigurablePassInterface.php]: I had to define it as a static method as I need to get an analyzer's name before they are instantiated (see the Factory::getExpressionPasses method)

**Comment by K-Phoen** [comment]
I refactored a bit the configuration system to use a different approach.

Now, all analyzers are and have to be configurable. They also have to implement a new method `getMetadata` that will return an object that describes them.

These metadata are used for several things:
* to configure them (some analyzers might need options)
* to describe them : a description in english can be nice to explain to the user what the analyzer actually does
* to determine if an analyzer is enabled or not (they can even be disabled if the PHP version isn't compatible → that would fix #123)

If you're satisfied with the current implementation, I'll write some test and I'll probably make another PR to provide a description for each analyzer currently implemented.

**Comment by ddmler** [comment]
So where can the user then view the description?

For later PRs:
I think we can then edit the deprecatedFunctions analyzer for example and add a check for the php version config. So we don't show deprecated things for 7.0 when the user wants to run his things on 5.6 or something. :+1: 

Also we need to rename a few analyzers for example MagicMethod\GetParameterCheck ... at least we should remove get because the thing checks set and clone too :smile: 

And the notice message tag should not be "array.short-syntax" but the key used in config "array_short_definition" for all analyzers.

**Comment by K-Phoen** [comment]
> So where can the user then view the description?

For now, the only place is in the output generated by the `./bin/phpsa config:dump-reference` command.

> I think we can then edit the deprecatedFunctions analyzer for example and add a check for the php version config. So we don't show deprecated things for 7.0 when the user wants to run his things on 5.6 or something. 

With this PR, that's not possible with a single analyzer. An analyzer can be enabled or disabled for a given PHP version, but once this decision is made the analyzer hasn't access to the PHP version targeted by the user.
I see several solutions to this problem:
* creating several analyzers
* giving access to a "runtime configuration" to the analyzers. This "runtime configuration" would for instance contain the targeted PHP version, a list of installed extensions, …

I tend to prefer the first solution as it's more simple to implement.

**Comment by ddmler** [comment]
@K-Phoen 
But "Context" is injected into every analyzer. You can get with `$context->application->configuration` the Configuration object so it should be possible or am I wrong?

**Comment by K-Phoen** [comment]
You're right, but to me, that looks more like a coincidence ;)
In an analyzer, I expect the context to hold compilation-related context such as the variable accessible from the current scope, this kind of stuff. Accessing the global configuration through it seem a bit weird to me (and it breaks Demeter's law btw).

**Comment by K-Phoen** [comment]
@ddmler and @ovr, I think that we should merge this PR now. I'll make separate PR for the remaining tasks. For instance:
* disabling/enabling notices by levels like CHECK_SAFE/CHECK_BETA/…
* improving documentation regarding analyzers configuration
* improving notices generated by analyzers to include a better description and a link to a page (in the wiki probably)
* …

WDYT?

**Comment by ddmler** [comment]
yeah we can do that, but i merged the exit analyzer so we have a conflict again. For descriptions of analyzers we have a documentation page you could copy those texts and edit them if needed.

**Pull request merged**